collectors/CMPCollector.js

```diff
@@ -258,6 +308,41 @@ class CMPCollector extends BaseCollector {
         }
     }
 
+    /**
+     * @returns {Promise<string[]>}
+     */
+    async scanPatterns() {
+        /**
+         * @type {string[]}
+         */
+        const found = [];
+        const pages = await this.context.pages();
+        if (pages.length > 0) {
+            const page = pages[0];
+            /**
+             * @type {Promise<string>[]}
+             */
+            const promises = [];
+            page.frames().forEach(frame => {
```

Comment:
`innerText` does exclude invisible text, as opposed to `textContent`. It would be nicer to capture shadow DOM though, since some CMPs use it. I'll look into those methods, thanks for sharing!

Reply:
Those particular methods are not ideal I'm afraid. Apart from more complicated usage,
- `performSearch` works with substrings, so we can't use regex with it. It's cumbersome to collect all possible substrings that may be matched by a simple regexp..
- `DOMSnapshot` cannot tell apart visible and invisible text. This is pretty important in this case, as CMPs often hide popups with `display: none`, but are still present on the page

I think `innerText` is still better because it's simple and works well for this purpose. For (open) shadow DOM, we can just [use JS](https://github.com/puppeteer/puppeteer/blob/main/src/common/QueryHandler.ts#L175), I can do it in a follow up PR (as part of [this task](https://app.asana.com/0/0/1202635343225979/1202986922914755/f))

Reply:
Gotcha, thank you for testing those out. I'm not sure if we can get the insides of the closed shadow dom via JS, but we should be able to at least get open shadow dom. Let's keep that for the follow up.